pkg/controller/servicebindingrequest/bindingdata.go

```diff
@@ -0,0 +1,162 @@
+package servicebindingrequest
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/redhat-developer/service-binding-operator/pkg/converter"
+	"github.com/redhat-developer/service-binding-operator/pkg/log"
+)
+
+// BindingDataHandler represents the binding data accumulated from backing
+// services, and later persisted as a secret or a configMap.
+type BindingDataHandler struct {
+	logger *log.Log          // logger instance
+	client dynamic.Interface // Kubernetes API client
+	plan   *Plan             // plan instance
+}
+
+// customEnvParser parse informed data in order to interpolate with values provided by custom
+// environment component.
+func (s *BindingDataHandler) customEnvParser(data map[string][]byte) (map[string][]byte, error) {
+	// transforming input into format expected by custom environment parser
+	cache := make(map[string]interface{})
+	for k, v := range data {
+		cache[k] = v
+	}
+
+	// interpolating custom environment
+	envParser := NewCustomEnvParser(s.plan.SBR.Spec.CustomEnvVar, cache)
+	values, err := envParser.Parse()
+	if err != nil {
+		return nil, err
+	}
+
+	for k, v := range values {
+		data[k] = []byte(v.(string))
+	}
+	return data, nil
+}
+
+// buildResourceClient creates a resource client to handle corev1/secret resource.
+func (s *BindingDataHandler) buildResourceClient(resource string) dynamic.ResourceInterface {
+	gvr := corev1.SchemeGroupVersion.WithResource(resource)
+	return s.client.Resource(gvr).Namespace(s.plan.Ns)
+}
+
+// createOrUpdate will take informed payload and create/update the binding secret/configmap
+// one.
+// If the spec.BindingReference is nil,
+// Returns an error when the Kubernetes client returns an error.
+func (s *BindingDataHandler) createOrUpdate(payload map[string][]byte) (*unstructured.Unstructured, error) {
+	ns := s.plan.Ns
+	name := s.plan.Name
+	logger := s.logger.WithValues("Namespace", ns, "Name", name)
+
+	var bindingRef interface{}
+
+	// Default binding resource type is Secret
+	resource := SecretResource
```

Comment:
this assigning `resource := SecretResource` should be after the `s.plan.SBR.Spec.BindingReference.ObjectType.Kind == ConfigMapKind` check in an `else`. So currently everytime secret is being created even before checking.

Reply:
Yeah, the flow is more like "Start with secret, and continue with secret unless specified otherwise"

1) Assume, we need to have a `Secret`

```
// Default binding resource type is Secret
resource := SecretResource
gvk := corev1.SchemeGroupVersion.WithKind(SecretKind)
```

2) If something is explicitly specified, and that something is a `configmap`, we'll go for configmap. Therefore, if someone specifies ObjectType.Kind = Service, we'll ignore it and create a binding secret.
```
// Only if otherwise specified as configmap, we'll use a binding
// configmaps
if s.plan.SBR.Spec.Binding != nil &&
	s.plan.SBR.Spec.Binding.ObjectType.Kind == ConfigMapKind {
	gvk = corev1.SchemeGroupVersion.WithKind(s.plan.SBR.Spec.Binding.ObjectType.Kind)
		resource = ConfigMapResource
}
```